src/osd/PG.cc

```diff
@@ -4484,6 +4484,29 @@ bool PG::may_need_replay(const OSDMapRef osdmap) const
   return crashed;
 }
 
+void PG::check_full_transition(OSDMapRef lastmap, OSDMapRef osdmap)
+{
+  if (osdmap->test_flag(CEPH_OSDMAP_FULL) &&
+      !lastmap->test_flag(CEPH_OSDMAP_FULL)) {
+    dout(10) << " cluster was marked full in " << osdmap->get_epoch()
+	     << dendl;
+    info.history.last_epoch_marked_full = osdmap->get_epoch();
```

Comment:
if we want last_epoch_marked_full to stay the beginning of the full period, we need to check for both pool and cluster full before resetting last_epoch_marked_full, or the pool and cluster full epochs could overwrite each other